### PR TITLE
LRDOCS-4117, writing XML workflow definitions

### PR DESCRIPTION
**Pull request by rbohl**
https://issues.liferay.com/browse/LRDOCS-4117, articles for writing XML workflow deifnitions. Let me know what you think about the structure of this, and whether they are complete enough or not. This is based on master, our branch for 7.1 docs, but could be backported for 7.0 as well.

### CODE REVIEW

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+ 
+In the JAR file, navigate to 
+
+    META-INF/definitions/
+
+and extract the four XML workflow definition files. These definitions provide
+good reference material for many of the workflow features and elements described
+in these articles. In fact, most of the XML snippets you see are lifted
+directly from these definitions.
+
```
Comment (natocesarrego) [inline develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown:40]: Hey @rbohl, after some interactions with Support/customers I've realized that good part of the Workflow's users are not technical, hence the steps described above could be are little complicated. In order to make easier to our non-tech audience access these embedded definitions, what do you think about attach them to this article? Once the definitions are visible to CE users, I understand that wouldn't be a problem publish it in the dev.liferay.com.

What do you think about it?

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+### Schema
+
+The XML structure of a workflow definition is defined in an XSD file,
+
+    liferay-worklow-definition-7_0_0.xsd
+
+Declare the schema at the top of the workflow definition file:
+
+    <?xml version="1.0"?
```
Comment (natocesarrego) [inline develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown]: A **>** is missing at the end of **<?xml version="1.0"?**. 😉

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+Give the definition a name, description, and version:
+
+    <name>Category Specific Approval</name>
+    <description>A single approver can approve a workflow content.</description>
+	<version>1</version>
+
+### Nodes
+
+After your definition has its schema and metadata defined, tackle the guts of
```
Comment (natocesarrego) [inline develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown]: What is the meaning of the expression **tackle the guts**? From what I can see, it's an expression equivalent to **face the problem**, am I right?

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+the workflow definition. *Nodes*, with their sub-elements, are the fundamental
+building blocks making up workflow definitions.
+
+- State Nodes don't require user input. The workflow does whatever is specified
+    in the state node's `actions` tag (a notification and/or a custom script) ,
```
Comment (natocesarrego) [inline develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown]: Extra blank space after parentheses and before comma.

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown [hunk 5]
```diff
@@ -0,0 +1,275 @@
+# Tasks
+
+Task nodes are fundamental parts of a workflow definition. When you start
+defining your organization's business processes and planning workflows, you
+likely first envision the tasks. As the name implies, tasks are the part of the
+workflow where *work* must be done. A user enters the picture and has to
+interact with the submitted asset. This user often takes the role of reviewer,
+where an asset is sent to them and they must decide if it is acceptable,
```
Comment (natocesarrego) [inline develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown]: The pronouns **them** and **they** are referring the user, hence would be better replace them by **him** and **he** respectively.

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        </task-timer>
+    </task-timers>
+
+The above task timer creates a notification. Specify a time period in the
+`<delay>` tag, and specify what action to take after the delay is up in the
+`<timer-actions>` block. The `<blocking>` element is used to specify whether the
+timer actions may recur. If `<blocking>false</blocking>`, make the timer action
+recur using the `<recurrence>` tag:
+
+    <recurrence>
```
Comment (natocesarrego) [inline develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown]: The **recurrence** tag should be nested to the **task-timer** tag? I was unsure about that during the reading. 😞

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown [hunk 7]
```diff
@@ -59,242 +54,14 @@ XML (it's practically poetic), check out the XSD
...
-                    <name>reject</name>
-                    <target>update</target>
-                    <default>false</default>
-                </transition>
-            </transitions>
-        </task>
-
-To learn more about workflow tasks, read the next article.
+Once the schema and metadata are in place, it's time to turn up the funky beats
+and get into the flow (the workflow). Learn about workflow nodes in the next
```
Comment (natocesarrego) [inline develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown]: Nice sentence! I liked that. 😄

**Review by natocesarrego — COMMENTED** [review]
Hi @rbohl,

I've taken more time than usual to review this doc due to its size and details, but now it's fully reviewed. Besides, congratulations for this article! In my opinion, it's simple and straightforward. I'm very, very happy to see the Workflow documentation growing and getting better. 😄 👏 🎉

Best regards.

**Comment by natocesarrego** [comment]
Hey @rbohl,

There are some comments marked as outdated, but they still make sense. Please take a look on them too.

Thanks in advance.

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown (hunk 1 above)
Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown:40]: @natocesarrego I think it's a good idea. Can you test this download link? It should download a ZIP with the XML files from dev.liferay.com: https://dev.liferay.com/documents/10184/1530512/Workflow+Definitions+Zip/60504133-f6ab-7587-f734-6275749fe450

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown (hunk 2 above)
Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown]: Thanks! I'll fix it.

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown (hunk 3 above)
Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown]: "Tackle the guts" is like saying, "work on the main part of". I'll change it to sound less confusing for non-English speakers.

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown (hunk 7 above)
Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown]: Thanks!

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown (hunk 5 above)
Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown]: @natocesarrego how about this sentence instead?

*This user often takes the role of reviewer, deciding if an asset they've received in the workflow is acceptable for publication in its current state, or perhaps needs more work.*

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown (hunk 6 above)
Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown]: Yeah, it's nested below `<task-timer>`, at the same level as `<delay>`. I'll make this more clear.

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown (hunk 1 above)
Comment (natocesarrego) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown:40]: Yes, I can and I could download it successfully. However, the definitions available on this ZIP are from 6.2.

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown (hunk 5 above)
Comment (natocesarrego) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown]: Hey @rbohl,

The both sentences are good to me, I was just concerned if the pronoun **they** is referring the single mentioned user with the reviewer role or the assets being reviewed. Is it clearer now? Sorry about that. 😬

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown (hunk 6 above)
Comment (natocesarrego) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown]: Nice! Thanks. 😄

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown (hunk 1 above)
Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown:40]: @natocesarrego Unfortunately that's what we ship with portal. I extracted these from a 7.0 CE GA5 bundle. Maybe I should go manually update the schema definition and re-upload them, so customers aren't confused.

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown (hunk 5 above)
Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/01-tasks.markdown]: @natocesarrego No, it was clear what you meant, but when you get me to review my own writing again, a lot of times I think of a better way to say the same thing. So as long as you think this is good, we'll stick with it. :)

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown (hunk 1 above)
Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown:40]: Updated the zip.

Comment (rbohl) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown:40]: @natocesarrego Should I still retain the instructions for extracting them manually? Like this:

> Only one workflow definition is installed by default: Single Approver. However,
> several more are embedded in the source code of your @product@ installation. If
> you're comfortable extracting the XML files from a JAR file embedded in an LPKG
> file, you're welcome to follow the steps below to obtain the workflow
> definitions. To obtain the files more conveniently, download a ZIP file
> [here](https://dev.liferay.com/documents/10184/1530512/Workflow+Definitions+Zip/60504133-f6ab-7587-f734-6275749fe450).
> 
> To extract the definitions for yourself, navigate to ...

**Comment by rbohl** [comment]
@natocesarrego check out my last commit. It addresses your review comments and adds a few of my own edits. Let me know if you think this is ready.

File: develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown (hunk 1 above)
Comment (natocesarrego) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown:40]: Hi @rbohl, if you've extracted it from the 7.0 bundle, then we should fix it in the code. 😉 I'll check it and talk to the Workflow team.

Regarding the instructions to get the definition manually, I understand that if we're attaching them to the article, then these instructions are not necessary anymore.

What do you think?

Comment (natocesarrego) [reply develop/tutorials/articles/150-workflow/01-creating-workflow-definitions/00-creating-workflow-definitions.markdown:40]: Hey @rbohl, on the other hand, I was checking your new commit and I think that could be a good option offer the two possibilities, link to download and instructions to extract the ZIP, for the readers get the Workflow definitions. Anyway, I'll let you take this decision. 😉

**Comment by natocesarrego** [comment]
Green light for me, @rbohl! 😄

**Comment by rbohl** [comment]
Forwarded here: https://github.com/sez11a/liferay-docs/pull/2835

**Pull request closed without merging**